pkg/plugin/healthz.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft and contributors.  All rights reserved.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"google.golang.org/grpc"
+	pb "k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1"
+	"k8s.io/klog/v2"
+
+	"github.com/Azure/kubernetes-kms/pkg/version"
+)
+
+const (
+	healthCheckPlainText = "healthcheck"
+)
+
+type HealthZ struct {
+	KMSServer      *KeyManagementServiceServer
+	HealthCheckURL *url.URL
+	UnixSocketPath string
+	RPCTimeout     time.Duration
+}
+
+// Serve creates the http handler for serving health requests
+func (h *HealthZ) Serve() {
+	serveMux := http.NewServeMux()
+	serveMux.HandleFunc(h.HealthCheckURL.EscapedPath(), h.ServeHTTP)
+	if err := http.ListenAndServe(h.HealthCheckURL.Host, serveMux); err != nil && err != http.ErrServerClosed {
+		klog.Fatalf("failed to start health check server, err: %+v", err)
+	}
+}
+
+func (h *HealthZ) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	klog.V(5).Infof("Started health check")
+	ctx, cancel := context.WithTimeout(context.Background(), h.RPCTimeout)
+	defer cancel()
+
+	conn, err := h.dialUnixSocket()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+	defer conn.Close()
+
+	// create the kms client
+	kmsClient := pb.NewKeyManagementServiceClient(conn)
+	// check version response against KMS-Plugin's gRPC endpoint.
+	err = h.checkRPC(ctx, kmsClient)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+	// check the configured keyvault, key, key version and permissions are still
+	// valid to encrypt and decrypt with test data.
+	enc, err := h.KMSServer.Encrypt(ctx, &pb.EncryptRequest{Plain: []byte(healthCheckPlainText)})
```

Comment:
If this is actually making a roundtrip call to encrypt and decrypt, could these calls impact the api limits for key vault? I see `periodSeconds: 10` below, what should be a minimum required value such that this probe won't add to the impact it has on the limit. 

At a high level, this probe is testing the following: 

1. this kms plugin is up and running, ready to serve request
2. access against the configured kv is valid
3. key configured is valid

I'm curious if this is what other plugins are doing as well. Google seems to do the same. AWS seems to just do 1.

Reply:
The total # of health check rtt calls will be:
- liveness probe: 2 * 6 = 12 (every 10s)
- API server ping check: 2* 3  = 6 (every 20s)

Total calls should be 18 per min. The reason for using 10s is so the kms plugin is restarted after 3 failures (30s). API server will start failing if the kms health check ping test fails. In terms of limits, key operation allow per 10s are 1000 for HSM and 2000 otherwise. 

https://docs.microsoft.com/en-us/azure/key-vault/general/service-limits

This should be well within the limits. We can try to cache the health check result across multiple calls and serve that if it's most recent but this is more the current state of the world.

Reply:
Ok thanks for providing these numbers. Let's definitely test and verify we are well within these limits when we run load tests.

Reply:
sounds good!